hera_mc/cm_handling.py

```diff
@@ -358,144 +330,73 @@ def get_connection_dossier(self, hpn_list, rev, port, at_date, exact_match=False
                                                             conn.up_part_rev,
                                                             conn.upstream_output_port,
                                                             conn.start_gpstime)
-                        connection_dossier['connections'][ckey] = copy.copy(conn)
-                        up_parts.append(ckey)
-                if len(up_parts) == 0:
-                    up_parts = [PC.no_connection_designator]
-                if len(down_parts) == 0:
-                    down_parts = [PC.no_connection_designator]
-                up_parts.sort()
-                down_parts.sort()
-                if len(up_parts) > len(down_parts):
-                    down_parts = (down_parts + len(up_parts) * [down_parts[-1]])[:len(up_parts)]
-                elif len(down_parts) > len(up_parts):
-                    up_parts = (up_parts + len(down_parts) * [up_parts[-1]])[:len(down_parts)]
-                connection_dossier['ordered-pairs'].append([up_parts, down_parts])
+                        connection_dossier['conn'][prkey]['up'][ckey] = copy.copy(conn)
+
+                # Pair upstream/downstream ports for this part
+                port_type = {'up': 'downstream_input_port', 'down': 'upstream_output_port'}
+                for sk in connection_dossier['conn'][prkey]['paired']:
+                    pkey_tmp = {}
+                    for i, ck in enumerate(connection_dossier['conn'][prkey][sk]):
+                        po = getattr(connection_dossier['conn'][prkey][sk][ck], port_type[sk])
+                        pkey_tmp[po + '{:03d}'.format(i)] = ck
+                    connection_dossier['conn'][prkey]['paired'][sk] = [pkey_tmp[x] for x in sorted(pkey_tmp.keys())]
+                pad = len(connection_dossier['conn'][prkey]['paired']['down']) - len(connection_dossier['conn'][prkey]['paired']['up'])
+                if pad < 0:
+                    connection_dossier['conn'][prkey]['paired']['down'].extend([None] * abs(pad))
+                elif pad > 0:
+                    connection_dossier['conn'][prkey]['paired']['up'].extend([None] * abs(pad))
```

Comment:
I'm having some trouble understanding the logic of this stanza and convincing myself that the sort algorithm will pair all of the connections correctly. Some explanatory comments might help.

Also, in the function's docstring, I think it would be very helpful to paste an example dictionary that you might get out of this function:

```python
""".... An example return value might look like:

{
  'up': {
    'blah': <Connection object>,
    'other: <Connection object>,
  },
  'down': {
     'yadda': ...
  },
  'paired': {
     'up': [
        'key000',
        ...
      ],
      'down': [
         None,
         ....,
      ]
   }
}
"""
```

Also also, these dossier dictionaries are getting big and complicated enough that they should really become classes. That way their fields can be documented and they can be given methods for manipulating their contents. But that is not a change to undertake right now.

Reply:
I COMPLETELY agree about the Class issue.  Tempting to do now, but better for another PR, which I will likely start.